File: zig/zigbeam/src/skynet.zig
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until depth_max levels deep (10^depth_max =
// 100M leaf tasks) and sums their results. Uses zigbeam's Loom work-stealing
// pool: the 10-way fan-out is expressed as a recursive binary split via
// loom.joinOnPool (Rayon-style fork-join with steal-while-wait), the idiomatic
// divide-and-conquer shape for this library. The whole computation runs as a
// pool task so the requested thread count does all the work while main parks.
const std = @import("std");
const loom = @import("loom");
const bu = @import("benchutil");

const depth_max = 8;
const iter_count = 1;

var pool: *loom.ThreadPool = undefined;
var done: bu.DoneEvent = .{};

fn skynetOne(base: u64, depth: u32) u64 {
    if (depth == depth_max) return base;
    var offset: u64 = 1;
    var i: u32 = 0;
    while (i < depth_max - depth - 1) : (i += 1) {
        offset *= 10;
    }
    return skyRange(base, depth, offset, 0, 10);
}

/// Sum children [lo, hi) of a node by recursive binary fork-join.
fn skyRange(base: u64, depth: u32, offset: u64, lo: u64, hi: u64) u64 {
    if (hi - lo == 1) return skynetOne(base + offset * lo, depth + 1);
    const mid = lo + (hi - lo) / 2;
    const l, const r = loom.joinOnPool(
        pool,
        skyRange,
        .{ base, depth, offset, lo, mid },
        skyRange,
        .{ base, depth, offset, mid, hi },
    );
    return l + r;
}

const RootCtx = struct {
    result: u64 = 0,

    fn run(ctx_opaque: *anyopaque) void {
        const ctx: *RootCtx = @ptrCast(@alignCast(ctx_opaque));
        ctx.result = skynetOne(0, 0);
        done.set();
    }
};

fn skynet(expected: u64) void {
    var ctx = RootCtx{};
    var task = loom.Task{ .execute = RootCtx.run, .context = @ptrCast(&ctx), .scope = null };
    done.reset();
    pool.spawn(&task) catch @panic("spawn failed");
    done.wait();
    if (ctx.result != expected) {
        bu.print("ERROR: wrong result - {d}\n", .{ctx.result});
    }
}

pub fn main(init: std.process.Init.Minimal) void {
    const argv = bu.Argv(1).init(init.args);
    const thread_count = argv.int(u32, 0) orelse bu.defaultThreads();
    bu.print("threads: {d}\n", .{thread_count});
    // Steal-while-wait joins nest stolen subtrees on the waiting worker's
    // stack, so recursive fork-join workloads need far more than the default
    // thread stack. Virtual memory: only touched pages are committed.
    pool = loom.ThreadPool.init(bu.allocator, .{
        .num_threads = thread_count,
        .stack_size = 1 << 30,
    }) catch @panic("pool init failed");

    var leaves: u64 = 1;
    var i: u32 = 0;
    while (i < depth_max) : (i += 1) {
        leaves *= 10;
    }
    const expected = (leaves - 1) * leaves / 2;

    skynet(expected); // warmup

    bu.print("runs:\n", .{});
    const start = bu.nowNs();
    var iter: usize = 0;
    while (iter < iter_count) : (iter += 1) {
        skynet(expected);
    }
    const dur_us = (bu.nowNs() - start) / std.time.ns_per_us;
    bu.print("  - iteration_count: {d}\n", .{iter_count});
    bu.print("    duration: {d} us\n", .{dur_us});
    bu.print("    max_rss: {d} KiB\n", .{bu.peakMemoryUsageKiB()});

    pool.deinit();
}

File: rust/tokio/src/bin/skynet.rs
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until DEPTH_MAX levels deep (10^DEPTH_MAX = 100M
// leaf tasks) and sums their results. Here each child is a `tokio::spawn`, so
// this measures tokio's task spawn/join throughput under deep nested fan-out.

use std::time::Instant;
use tokio_bench::{multi_thread_runtime, peak_memory_usage, BoxFut};

const DEPTH_MAX: u64 = 8;
const ITER_COUNT: usize = 1;

fn skynet_one(base: u64, depth: u64) -> BoxFut<u64> {
    Box::pin(async move {
        if depth == DEPTH_MAX {
            return base;
        }
        let mut offset = 1u64;
        for _ in 0..(DEPTH_MAX - depth - 1) {
            offset *= 10;
        }

        let mut handles = Vec::with_capacity(10);
        for i in 0..10u64 {
            handles.push(tokio::spawn(skynet_one(base + offset * i, depth + 1)));
        }

        let mut count = 0u64;
        for h in handles {
            count += h.await.unwrap();
        }
        count
    })
}

async fn skynet(expected: u64) {
    let count = skynet_one(0, 0).await;
    if count != expected {
        println!("ERROR: wrong result - {}", count);
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        tokio_bench::default_threads()
    };

    let leaves = 10u64.pow(DEPTH_MAX as u32);
    let expected = (leaves - 1) * leaves / 2;

    let rt = multi_thread_runtime(thread_count);
    println!("threads: {}", thread_count);

    rt.block_on(async move {
        skynet(expected).await; // warmup

        let start = Instant::now();
        for _ in 0..ITER_COUNT {
            skynet(expected).await;
        }
        let dur = start.elapsed();

        println!("runs:");
        println!("  - iteration_count: {}", ITER_COUNT);
        println!("    duration: {} us", dur.as_micros());
        println!("    max_rss: {} KiB", peak_memory_usage());
    });
}
